### DISCUSSION TITLE
LittleFS -> I2S | How do I trigger looping for specific files?

### DISCUSSION THREAD

**Discussion by auyiteng** [pschatzmann/arduino-audio-tools#914] [Q&A] [answered]
`/**
 * @file player-littlefs-i2s.ino
 * @brief example using the LittleFS library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  i2s.begin(cfg);

  // setup player
  //source.setFileFilter("*Bob Dylan*");
  player.setMetadataCallback(printMetaData);
  player.begin();
}

void loop() {
  player.copy();
}`

Hi, I want to try working on a project that is similar to this [https://www.youtube.com/watch?v=3YmMNpbFjp0](url)
I will have EASY button to activate easy.mp3 loop, PANIC button to activate panic.mp3 loop, BULLSHIT button to activate bullshit.mp3 loop, WTF button to activate wtf.mp3 loop, and a stop button to stop.

I've read the documentation for AudioPlayer [https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html#af62138cf6d06e41ae7d0535e4ecdeeb2](url) and I don't understand

1. how it's already playing through all the files when I haven't called player.play()?
2. how do i loop it? is there no native function like player.loop() to loop through the current mp3 file?
3. how does the filefilter work? am I able to change during runtime, since its in setup()?

**Comment by auyiteng** [comment]
I found the [setLoop](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_memory_stream.html#a808d67693c38dccdc6b4692c65d6c36f) setLoop function along with MemoryStream, but how do I integrate that with my LittleFS setup, and how do I change the file being played with filefilter for different buttons?

**Comment by pschatzmann** [comment]
Please read the Wiki about the AudioPlayer: From my point of view it contains all the information and I don't think there is anything to add... 
ps. you dont need to use the AudioPlayer: it should be also quite trivial to build something based on the FileLoop class if you insist on using files

**Comment by auyiteng** [comment]
I'm assuming you referred to this page [https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class#file-based-audio-sources](url) with the class for littleFS [https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source_little_f_s.html](url) I don't see any method to loop it.

I'm looking at the https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_file_loop.html file loop class right now but how to I pass it along into the AudioPlayer player variable I've defined?

**Comment by pschatzmann** [comment]
Here is the relevant information:
An [AudioSource](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source.html) is the abstract API used by the AudioPlayer class to access the audio data and to navigate. You can create your own source or just use one of the existing implementations. and [AudioSource Callback](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source_callback.html) Implement your own logic just by providing some simple callback methods

**Comment by auyiteng** [comment]
Just so I understand this right: AudioPlayer is the main class that's doing all the work, which the constructor takes in an AudioSource class(LittleFS), AudioOutput class(I2S), and AudioDecoder class(MP3).

I'm sorry I still don't understand how can I insert FileLoop into the equation. I tried finding for an example with FileLoop in the /examples directory but I couldn't seem to find one.

I've spent a couple of hours trying other different methods (calling player.begin() again after a set delay which kinda works??? but not ideal), and looking at past discussions

https://github.com/pschatzmann/arduino-audio-tools/discussions/724
(this somehow describes using MemoryStream to loop too but how?? the AudioPlayer constructor doesn't seem to take in a MemoryStream either)





And also I'm confused about how this example works:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd-audiokit/streams-sd-audiokit.ino

There the AudioPlayer.copy() method is not called in the loop but just a simple check is done, how is it still able to play the full mp3 file? 

void loop(){
  if (!player.copy()) {
    stop();
  }
}

The documentation (https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html) here states that I have to call the copy() method inside the loop for the AudioPlayer class to work properly.

Thanks for reading and answering me earlier, this library is cool. Is it possible you can help to write one line of how FileLoop can be passed to the AudioPlayer class it would be really helpful for me, just the passing method, how do I pass the FileLoop to AudioPlayer. Thank you!

**Reply by auyiteng** [reply]
I would really want to try to get FileLoop to work as it seems like its what I want

**Reply by LinusHeu** [reply]
> There the AudioPlayer.copy() method is not called in the loop but just a simple check is done, how is it still able to play the full mp3 file?

This example doesn't use the AudioPlayer class. Instead, it uses the streams and the copier. (If you look into the code of the AudioPlayer class, you will see, that it uses the streams internally as well.)
And the copier IS called. `if(!player.copy())` calls player.copy(). And player.copy() returns a boolean (or something that can be converted to a boolean). Based on that return value, it goes into the if (stop()) or skips the if.

You can do that with any method that returns a boolean or something like that.
Example:
```ino
bool isEven(int number){
  //this method can not only calculate its return value. It can also do other stuff like copy audio or say hello.
 Serial.println("Hello. I'm the Method isEven()");
 return number % 2; //modulo
}

void loop(){
  if(isEven(2)){
  Serial.println("it is even");
  }
}
```

**Comment by pschatzmann** [comment, marked as answer]
I am afraid you still don't understand: Just implement your own AudioSource logic! All you need to do is to implement the nextStream method to return a Stream (=File) that makes sense in your case (dependent on what button was pressed). 

You could return a FileLoop but this is not really necessary as when you return the same file, it is automatically playing again

However if I would implment your requirement, I would not use the player at all, but just build my own logic using directly multiple FileLoop objects as copy source. I would even replace the files with a MemoryStream to simplify things even further...

**Comment by auyiteng** [comment]
`/**
 * @file player-littlefs-i2s.ino
 * @brief example using the LittleFS library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  pinMode(4, INPUT_PULLUP)
  pinMode(2, INPUT_PULLUP);
  pinMode(5, INPUT_PULLUP);
  pinMode(17, INPUT_PULLUP);
  pinMode(32, INPUT_PULLUP);
  pinMode(33, INPUT_PULLUP);
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = 15;
  cfg.pin_ws = 12;
  cfg.pin_data = 14;
  i2s.begin(cfg);

  // setup player
  //source.setFileFilter("*Bob Dylan*");
  player.setMetadataCallback(printMetaData);
  //player.begin();
}

void loop() {
  player.copy();
  if(digitalRead(4) == LOW) {
    source.setFileFilter("*easy*");
    player.begin();  
  }
  else if(digitalRead(2) == LOW) {
    source.setFileFilter("*panic*");
    player.begin();  
  }
  else if(digitalRead(5) == LOW) {
    source.setFileFilter("*bullshit*");
    player.begin();  
  }
  else if(digitalRead(17) == LOW) {
    player.stop();  
  }
}
`


do you have any idea why on the serial monitor it spams:
[I] Fade.h : 104 - faded out 576 frames to volume 0.000000

**Comment by auyiteng** [comment]
Okay, it seems to spam that if player.begin() is called after player.stop()

Does player.stop() act as killing the AudioPlayer entirely and not just a graceful stop playing, but still allow for playing later on?

**Comment by auyiteng** [comment]
I'm able to replicate this using this bare minimum code

`/**
 * @file player-littlefs-i2s.ino
 * @brief example using the LittleFS library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"
#include "AudioCodecs/CodecMP3Helix.h"

enum {itsStopped, itsPlaying};
unsigned char itsState;

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  pinMode(4, INPUT_PULLUP);
  pinMode(2, INPUT_PULLUP);
  pinMode(5, INPUT_PULLUP);
  pinMode(17, INPUT_PULLUP);
  pinMode(32, INPUT_PULLUP);
  pinMode(33, INPUT_PULLUP);
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = 15;
  cfg.pin_ws = 12;
  cfg.pin_data = 14;
  i2s.begin(cfg);

  // setup player
  //source.setFileFilter("*Bob Dylan*");
  //player.setMetadataCallback(printMetaData);
  player.begin();
}

void loop() {
  player.copy();
  switch (itsState) {
    case itsStopped:
      Serial.println("itsStopped");
      if(digitalRead(2) == LOW) { //testingButtonPressed
        //player.stop();
        //source.setFileFilter("*test*");
        player.begin();
        itsState = itsPlaying;
        break;
      }
      else {
        break;
      }
    case itsPlaying:
      Serial.println("itsPlaying");
      if(digitalRead(33) == LOW) { //stopButtonPressed
        player.stop();
        itsState = itsStopped;
        break;
      }
      else {
        break;
      }
    default:
      itsState = itsStopped;
      break;
  }
}`

If I pull down GPIO 33 (player.stop()), then pull down GPIO2 (player.begin()), nothing gets played and instead my serial monitor is spammed with [I] Fade.h : 104 - faded out 576 frames to volume 0.000000

**Comment by pschatzmann** [comment]
Don't paste code w/o proper formatting: this is impossible to read!
I guess you just forgot to debounce your input! (e.g. by using the AudioActions or by using a pin interrupt instead of a digitalRead, or any other Debounce implememntation)

**Comment by auyiteng** [comment]
I apologise for the formatting issues. I tried this with a delay instead of pushbuttons to remove any possible debouncing and still encountered this issue. The following code executes player.begin() in the setup function, and alternates between player.stop() and player.begin() every 5 seconds in the loop function.

The program runs perfectly from 0 to 5 seconds, and stops correctly from 5 to 10 seconds, but at the 10 second mark, I still encounter the same issue with [I] Fade.h : 104 - faded out 576 frames to volume 0.000000 being spamming in my console and the audio jittering and not playing correctly when subsequent times player.begin() gets called after the first player.stop().

Please refer to the below minimal reproducible code, and here's a link to the pastebin of the serial monitor output: https://pastebin.com/69pFGcbc

```
#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"
#include "AudioCodecs/CodecMP3Helix.h"

unsigned long previousMillis = 0UL;
unsigned long interval = 1000UL;

enum {itsStopped, itsPlaying};
unsigned char itsState;

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = 15;
  cfg.pin_ws = 12;
  cfg.pin_data = 14;
  i2s.begin(cfg);

  player.begin();
  itsState = itsPlaying;
}

void loop() {
  player.copy(); //can't have any delay, activates StreamCopy.h

  unsigned long currentMillis = millis();
  if(currentMillis - previousMillis > 5000) //every 5s
  {
    Serial.println("5s check triggered");
    if (itsState == itsPlaying)
    {
        Serial.println("stop triggered");
        player.stop();
        itsState = itsStopped;
    }
    else {
        Serial.println("play triggered");
        player.begin();
        itsState = itsPlaying;
    }

  // Don't forget to update the previousMillis value
    previousMillis = currentMillis;
  }
} 
```

**Reply by auyiteng** [reply]
PS. Thank you for helping

**Reply by auyiteng** [reply]
Is this a bug?

**Comment by pschatzmann** [comment]
The opposite of stop() is play() and not begin()! Alternativly you can use setActive(false) and setActive(true);
You also don't need your state variable because you can query the state with player.isActive() so you can toggle the state with player.setActive(!player.isActive());

If fading is not important to you you can alternatively just stop and resume calling copy()

**Reply by auyiteng** [reply]
According to the documentation [AudioPlayer](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html)
it seems that begin() is used to start playing from the beginning again, which seems like what I want. play() on the other hand according to the documentation seems to resume playing from where the playhead was last when stop() was called

**Reply by auyiteng** [reply]
Does begin() not work after stop() is called?

**Comment by auyiteng** [comment]
Okay, I have rewrote the code to call play() instead of begin(). It seems to not like that begin() is being called more than once, maybe I misunderstood the documentation.

In any case, is there a way to replay from the beginning again? Do I declare 
AudioPlayer player(source, i2s, decoder);
again?

**Reply by pschatzmann** [reply]
did you try to call setIndex(0) ?

begin() is working as well, you just need to be aware that this might take more then the 5 seconds that you were granting...
You might be able to reduce this by optimizing the timouts settings in the AudioSource

**Reply by auyiteng** [reply]
Tysm! Calling player.setIndex(0), then calling player.play() successfully restarted playing from the beginnning.

For the purposes of looping when the AudioPlayer has finished playing the mp3, is there was a way to check if AudioPlayer reached the end of the file? I tried using isActive() but it always returns true even if the stream ended

```
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
5s check triggered
1
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 81 -> 81 -> 81 bytes - in 1 hops
5s check triggered
1
5s check triggered
1
```

**Reply by pschatzmann** [reply]
Just check the result of the copy call. I am not sure but getStream() might help as well: I dont remember if it returns nullptr or you can check if available() is 0

**Comment by auyiteng** [comment]
isActive() seems to only return false when stop() is explicitly called, even though StreamCopy.h reached the end of file and nothing is being played. Is there a way to check if AudioPlayer is not playing anything?

**Reply by pschatzmann** [reply]
See my answer above

**Comment by auyiteng** [comment]
Thank you
